tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp

```diff
@@ -0,0 +1,216 @@
+//===--- libSwiftSyntaxParser.cpp - C API for Swift Syntax Parsing --------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This C API is primarily intended to serve as the Swift parsing component
+// of SwiftSyntax (https://github.com/apple/swift-syntax).
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift-c/SyntaxParser/SwiftSyntaxParser.h"
+#include "swift/AST/Module.h"
+#include "swift/Basic/LangOptions.h"
+#include "swift/Basic/SourceManager.h"
+#include "swift/Parse/Parser.h"
+#include "swift/Parse/SyntaxParseActions.h"
+#include "swift/Syntax/Serialization/SyntaxSerialization.h"
+#include "swift/Subsystems.h"
+#include <Block.h>
+
+using namespace swift;
+using namespace swift::syntax;
+using namespace swift::byteTree;
+
+typedef swiftparse_range_t CRange;
+typedef swiftparse_client_node_t CClientNode;
+typedef swiftparse_syntax_node_t CRawSyntaxNode;
+typedef swiftparse_trivia_piece_t CTriviaPiece;
+typedef swiftparse_syntax_kind_t CSyntaxKind;
+
+namespace {
+class SynParser {
+  swiftparse_node_handler_t NodeHandler = nullptr;
+  swiftparse_node_lookup_t NodeLookup = nullptr;
+
+public:
+  void setNodeHandler(swiftparse_node_handler_t hdl) {
+    auto prevBlk = NodeHandler;
+    NodeHandler = Block_copy(hdl);
+    Block_release(prevBlk);
+  }
+
+  void setNodeLookup(swiftparse_node_lookup_t lookupBlk) {
+    auto prevBlk = NodeLookup;
+    NodeLookup = Block_copy(lookupBlk);
+    Block_release(prevBlk);
+  }
+
+  ~SynParser() {
+    setNodeHandler(nullptr);
+    setNodeLookup(nullptr);
+  }
+
+  swiftparse_client_node_t parse(const char *source);
+};
+
+class CLibParseActions : public SyntaxParseActions {
+  SourceManager &SM;
+  unsigned BufferID;
+  swiftparse_node_handler_t NodeHandler;
+  swiftparse_node_lookup_t NodeLookup;
+
+public:
+  explicit CLibParseActions(SourceManager &sm, unsigned bufID,
+                            swiftparse_node_handler_t hdl,
+                            swiftparse_node_lookup_t lookup)
```

Comment:
Similarly, this constructor can take a `SynParser&` as the third parameter.

Reply:
Done.